### PR TITLE
[FAI-58] Extend DMNRuntimeEventListener

### PR DESCRIPTION
**Pull request by kostola**
Add methods + events to DMNRuntimeEventListener to be able to monitor when API method calls are invoked on the DMNRuntime (evaluateAll, evaluateDS, etc.)

### CODE REVIEW

File: kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/AfterEvaluateAllEvent.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package org.kie.dmn.api.core.event;
+
+public interface AfterEvaluateAllEvent extends DMNEvent {
```
Comment (kostola) [inline kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/AfterEvaluateAllEvent.java]: @tarilabs please check that the structure of `AfterEvaluateAllEvent` makes sense to you

File: kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/BeforeEvaluateAllEvent.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package org.kie.dmn.api.core.event;
+
+public interface BeforeEvaluateAllEvent extends DMNEvent {
```
Comment (kostola) [inline kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/BeforeEvaluateAllEvent.java]: @tarilabs please check that the structure of `BeforeEvaluateAllEvent` makes sense to you

File: kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/BeforeEvaluateAllEvent.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+    String getModelNamespace();
+
+    String getModelName();
+
+}
```
Comment (kostola) [inline kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/BeforeEvaluateAllEvent.java:25]: @tarilabs please check that the structure of `BeforeEvaluateAllEvent` makes sense to you

File: kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/AfterEvaluateAllEvent.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+    String getModelNamespace();
+
+    String getModelName();
+
+}
```
Comment (kostola) [inline kie-dmn/kie-dmn-api/src/main/java/org/kie/dmn/api/core/event/AfterEvaluateAllEvent.java:25]: @tarilabs please check that the structure of `AfterEvaluateAllEvent` makes sense to you

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java [hunk 5]
```diff
@@ -32,85 +47,114 @@
...
-    private static void notifyListeners(DMNRuntimeEventManager eventManager, Consumer<DMNRuntimeEventListener> consumer) {
+    public static AfterEvaluateAllEvent fireAfterEvaluateAll(DMNRuntimeEventManager eventManager, DMNModel model, DMNResult result) {
+        return notifyListeners(
+                eventManager,
+                () -> new AfterEvaluateAllEventImpl(model.getNamespace(), model.getName(), result),
+                DMNRuntimeEventListener::afterEvaluateAll
+        );
+    }
+
+    private static <T> T notifyListeners(DMNRuntimeEventManager eventManager, Supplier<T> eventSupplier, BiConsumer<DMNRuntimeEventListener, T> eventConsumer) {
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java]: I am missing why this change is needed, please?

Comment (tarilabs) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java]: Nevermind, I believe I got it, because to align with this: https://github.com/kiegroup/drools/commit/8116db88ff1acc028f306686ed02aaa658ea4e3f#diff-9e8ea9e754fcefdbad2891d584eaa112R34

Comment (tarilabs) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java]: please confirm.

**Comment by tarilabs** [comment]
Still considering this: https://github.com/kiegroup/drools/commit/8116db88ff1acc028f306686ed02aaa658ea4e3f#diff-9e8ea9e754fcefdbad2891d584eaa112R34
might be worthy to evaluate if isn't it better to use similar approach rather then using ID/UUID in the context?

**Comment by r00ta** [comment]
Any test? :)

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java (hunk 5 above)
Comment (r00ta) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java]: What about removing the Supplier with
```
    private static <T extends DMNEvent> T notifyListeners(DMNRuntimeEventManager eventManager, T event, BiConsumer<DMNRuntimeEventListener, T> eventConsumer) {
        if (!eventManager.hasListeners()) {
            return null;
        }

        for( DMNRuntimeEventListener listener : eventManager.getListeners() ) {
            try {
                eventConsumer.accept(listener, event);
```
?

Comment (danielezonca) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java]: Please verify performance implication too because I expect this new `notifyListeners` method might have a negative impact

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/AfterEvaluateAllEventImpl.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package org.kie.dmn.core.impl;
+
+import org.kie.dmn.api.core.DMNResult;
+import org.kie.dmn.api.core.event.AfterEvaluateAllEvent;
+
+public class AfterEvaluateAllEventImpl implements AfterEvaluateAllEvent {
+
+    private String modelNamespace;
+    private String modelName;
+    private DMNResult result;
```
Comment (danielezonca) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/AfterEvaluateAllEventImpl.java:26]: What about `final`?

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/BeforeEvaluateAllEventImpl.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package org.kie.dmn.core.impl;
+
+import org.kie.dmn.api.core.DMNResult;
+import org.kie.dmn.api.core.event.BeforeEvaluateAllEvent;
+
+public class BeforeEvaluateAllEventImpl implements BeforeEvaluateAllEvent {
+
+    private String modelNamespace;
+    private String modelName;
+    private DMNResult result;
```
Comment (danielezonca) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/BeforeEvaluateAllEventImpl.java:26]: What about `final`?

**Review by danielezonca — COMMENTED** [review]
Minor comments

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java (hunk 5 above)
Comment (kostola) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNRuntimeEventManagerUtils.java]: I just reverted it back to the original one just adding the two methods I need

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/AfterEvaluateAllEventImpl.java (hunk 6 above)
Comment (kostola) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/AfterEvaluateAllEventImpl.java:26]: I looked at classes implementing other DMN events and none of them has `final` private attributes, so I assumed there may be some reason I don't know yet.
I can mark them as `final` if there's no such reason :wink:

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/BeforeEvaluateAllEventImpl.java (hunk 7 above)
Comment (kostola) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/BeforeEvaluateAllEventImpl.java:26]: Same as [here](https://github.com/kiegroup/drools/pull/2802#discussion_r388801864)

**Comment by kostola** [comment]
@tarilabs @danielezonca @jiripetrlik this is ready for review now, please take a look :wink:

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNRuntimeListenerTest.java [hunk 8]
```diff
@@ -139,7 +142,11 @@ public void testListenerWithBKM() throws Exception {
         final DMNRuntime runtime = DMNRuntimeUtil.createRuntime("org/kie/dmn/core/say_for_hello.dmn", this.getClass());
         TestEventListener listener = new TestEventListener();
         runtime.addListener(listener);
-        final DMNModel dmnModel = runtime.getModel("http://www.trisotech.com/dmn/definitions/_b6f2a9ca-a246-4f27-896a-e8ef04ea439c", "say for hello");
+
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNRuntimeListenerTest.java]: wouldn't be easier to just have a separate test for the 2 new events?
you can even use an anonymous class listener just tailored to the 2 new events.
This way we guarantee non regression more easily

**Review by tarilabs — APPROVED** [review]
minor comment if you want to consider, the rest is fine thank you

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNRuntimeListenerTest.java (hunk 8 above)
Comment (kostola) [reply kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNRuntimeListenerTest.java]: I expanded the already existing test because I thought its purpose was to test the whole sequence of events, since it includes also `BeforeEvaluateDecisionEvent` and `AfterEvaluateDecisionEvent`.
I can easily move my logic to a separate one... @danielezonca what do you think?

Comment (danielezonca) [reply kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNRuntimeListenerTest.java]: It is the same for me, probably if we move this additional check to another test it could be easier to read

Comment (kostola) [reply kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNRuntimeListenerTest.java]: Ok I moved it in a separate test :+1:

**Review by danielezonca — APPROVED** [review]

**Review by jiripetrlik — APPROVED** [review]
Looks good to me.

**Comment by kostola** [comment]
@kiegroup/gatekeepers could you please merge this?
